### ISSUE TITLE
`__fish_macos_set_env` does not handle lines with colons

### ISSUE THREAD

**Issue by injust** [fish-shell/fish-shell#10684]
Using:
fish 3.7.1
macOS Sonoma 14.5
Ghostty terminal

---

Please help me understand what https://github.com/fish-shell/fish-shell/blob/master/share/config.fish#L190-L192 is doing, as its behaviour seems incorrect to me.

I use Ghostty terminal, which appends itself to MANPATH. fish inherits MANPATH as:

```
:/Applications/Ghostty.app/Contents/Resources/ghostty/../man
```

fish then runs https://github.com/fish-shell/fish-shell/blob/master/share/config.fish#L190-L192, which sets MANPATH to:

```
/usr/share/man:/usr/local/share/man::/Applications/Ghostty.app/Contents/Resources/ghostty/../man
```

That is, fish prepends the paths from `/etc/manpaths` to MANPATH. (My `/etc/manpaths.d/` is empty.)

`man manpath` says:

```
IMPLEMENTATION NOTES
     The manpath utility constructs the manual path from two sources:
     1.   From each component of the user's PATH for the first of:
          -   pathname/man
          -   pathname/MAN
          -   If pathname ends with /bin: pathname/../share/man and pathname/../man
     2.   The configuration files listed in the FILES section for MANPATH entries.
     The information from these locations is then concatenated together.

     If the -L flag is set, the manpath utility will search the configuration files listed in the FILES section for MANLOCALE entries.
```

My default `/etc/man.conf` already contains everything in `/etc/manpaths` (i.e. `/usr/share/man` and `/usr/local/share/man`), so fish is prepending paths that already exist in man's search path.

Effectively, fish is moving those two paths from 2) and putting them _before_ 1), but _only_ if MANPATH is non-empty. This means that the system man pages in these 2 paths override anything in your PATH. Why is fish doing this?

---

For context, some Homebrew packages shadow macOS built-ins, so Homebrew installs them under alternate names (e.g. `gsed` if you install [`gnu-sed`](https://formulae.brew.sh/formula/gnu-sed)). If you add `/usr/local/opt/gnu-sed/libexec/gnubin` to your PATH, then `sed` will run GNU sed, and `man sed` should show the man page for GNU sed. This doesn't work in fish, and I think this is why.

**Comment by faho** [comment]
As the comment says: This is what macOS' [path_helper](https://opensource.apple.com/source/shell_cmds/shell_cmds-203/path_helper/path_helper.c.auto.html) does in login shells, so we do the same thing.

**Comment by injust** [comment]
Another case of "should've read the comments"...

Sorry for the noise 😦

**Comment by injust** [comment]
> It seems weird to me that Ghosttty would set $MANPATH just to its own path, in systems I know this would override the system default manpath.

@faho I missed a `:` in my original post. Ghostty _appends_ itself to MANPATH.

**Comment by injust** [comment]
> As the comment says: This is what macOS' [path_helper](https://opensource.apple.com/source/shell_cmds/shell_cmds-203/path_helper/path_helper.c.auto.html) does in login shells, so we do the same thing.

`__fish_macos_set_env` leaves appended values as appended, but I noticed that `path_helper` converts appended values to _prepended_ values.

@faho Is `__fish_macos_set_env` supposed to mirror `path_helper` behaviour exactly? I'm not sure if this difference is intentional or not.

**Comment by faho** [comment]
> __fish_macos_set_env leaves appended values as appended, but I noticed that path_helper converts appended values to prepended values

The [source we have](https://opensource.apple.com/source/shell_cmds/shell_cmds-203/path_helper/path_helper.c.auto.html) as well as [this github mirror](https://github.com/apple-opensource-mirror/shell_cmds/blob/4ea11424b80a117ad1e401867b1453bb2362feb3/path_helper/path_helper.c#L4) appends.

Our code seems to be a fairly straightforward translation of that - read /etc/manpath, read files in /etc/manpaths.d/, read $MANPATH, append each directory not already included.

If you're saying that doesn't happen, I'm going to need some more information about that, because given the code I don't see a difference.

**Comment by injust** [comment]
If I'm reading the `path_helper` source correctly, `append_path_segment` [drops empty segments](https://github.com/apple-opensource-mirror/shell_cmds/blob/4ea11424b80a117ad1e401867b1453bb2362feb3/path_helper/path_helper.c#L45). So when processing Ghostty's `:/Applications/Ghostty.app/Contents/Resources/ghostty/../man`, the leading `:` is effectively dropped.

So it appends `/usr/share/man`, `/usr/local/share/man`, and then `/Applications/Ghostty.app/Contents/Resources/ghostty/../man`, resulting in:

```
/usr/share/man:/usr/local/share/man:/Applications/Ghostty.app/Contents/Resources/ghostty/../man:
```

This is what I meant by `path_helper` converting appended values to prepended values -- the order is preserved, but Ghostty's path gets moved before the trailing `:` (I'm actually not sure where the trailing `:` is being added).

---

Because fish processes the empty segments, you get (notice the `::` in the middle):

```
/usr/share/man:/usr/local/share/man::/Applications/Ghostty.app/Contents/Resources/ghostty/../man
```

This honestly makes more sense to me, but it just doesn't match `path_helper`.

I haven't tested, but there should be a similar discrepancy if `$MANPATH` contains a `::` in the middle.

**Comment by faho** [comment]
Our code also drops empty segments:

https://github.com/fish-shell/fish-shell/blob/325b51aca0974bd5409c6007e8e98b49a68f03c8/share/config.fish#L172-L173

**Comment by injust** [comment]
Oh, hmm. I have no idea then. Somehow I am getting

```/usr/share/man:/usr/local/share/man::/Applications/Ghostty.app/Contents/Resources/ghostty/../man```

and I have no idea why. zsh (which I believe actually runs `path_helper`) gives me

```
/usr/share/man:/usr/local/share/man:/Applications/Ghostty.app/Contents/Resources/ghostty/../man:
```

**Comment by faho** [comment]
Hold on, the entry is ":/Applications/Ghostty.app/Contents/Resources/ghostty/../man".

That's gonna be read as one line, and then it's gonna be added as one line, then fish's "path variable" magic turns it into two entries.

We expect the path files to have one path per line instead of the awkward ":" gunk, and this breaks that assumption.

I'm going to reopen this for someone on macos to write a patch that handles ":" in path lines because I can't test it.

**Comment by westernwontons** [comment]
I would like to take a jab at this if it’s unclaimed. I encountered this issue today when trying ghostty too. If it’s game, can I have a little guidance on where to start?

**Comment by faho** [comment]
Sure. In macos_set_env: https://github.com/fish-shell/fish-shell/blob/a579abb81bf883486f8d89181ce99ba291d84893/share/config.fish#L170-L175

You probably want to `string split :` *before* you check if the entry is already included.

**Comment by westernwontons** [comment]
Cheers!

**Comment by westernwontons** [comment]
This seems to work with Ghostty, although I'm not really confident about the nested while loops:

```fish
while read -l entry
    string split : $entry | while read -l split_entry
        if not contains -- $split_entry $result
            test -n "$split_entry"
            and set -a result $split_entry
        end
    end
end <$path_file
```

I tried passing `-d :` to `while`, but that wasn't doing what I expected it to do. With the nested while, all leading/trailing colons are removed. 

Funny thing tho that I noticed is that the behaviour between `fish` and `zsh` are inverted in my case compared to how @injust is experiencing it. Starting Ghostty with `command` set to `zsh --login -interactive` gives me:

```zsh
$ echo $MANPATH
/Users/redacted/.zi/polaris/man::/usr/share/man:/usr/local/share/man:/Applications/Ghostty.app/Contents/Resources/ghostty/../man
```

while for `fish`, I get the following:

```fish
$ echo $MANPATH
/usr/share/man /usr/local/share/man  /Applications/Ghostty.app/Contents/Resources/ghostty/../man
```

This revelation also made me find out that I was using `zsh` in Ghostty and not `fish` as I thought I was.

**Comment by faho** [comment]
>  although I'm not really confident about the nested while loops:

You don't need them if you redirect to `string`:

```fish
string split : <$path_file | while read
```

or use a for-loop instead:

```fish
for entry in (string split : <$path_file)
```

**Comment by westernwontons** [comment]
That for-loop sits better with my eyes imo. Here's the final version:

```fish
for path_file in $argv[2] $argv[3]/*
    if test -f $path_file
        for entry in (string split : <$path_file)
            if not contains -- $entry $result
                test -n "$entry"
                and set -a result $entry
            end
        end
    end
end
```

**Comment by faho** [comment]
> I'm using fish 4.0.0 now, which includes https://github.com/fish-shell/fish-shell/commit/95d61ea0fbd90ed01b6ea39a790ca0d1b6a14689

It doesn't. It appears we never committed it to the 4.0 branch.

Picking it for 4.0.1 now.

**Comment by injust** [comment]
> But the MANPATH behaviour hasn't changed and I'm still getting the same MANPATH as [#10684 (comment)](https://github.com/fish-shell/fish-shell/issues/10684#issuecomment-2564331523).

I'm using fish 4.0.2 (which hopefully includes the fix) but my MANPATH still doesn't seem correct.

**Comment by krobelus** [comment]
can you insert some debugging statements around `__fish_macos_set_env` in `share/config.fish` (check inputs, outputs etc.)

**Comment by injust** [comment]
On fish 4.1.2 now, with this added to the beginning and end of `__fish_macos_set_env`:

```fish
if test "$argv[1]" = MANPATH
    echo \$MANPATH = "$MANPATH"
end
```

outputs:

```
$MANPATH = :/Applications/Ghostty.app/Contents/Resources/ghostty/../man
$MANPATH = /usr/share/man:/usr/local/share/man::/Applications/Ghostty.app/Contents/Resources/ghostty/../man
```

---

With completely unconfigured zsh, I get:

```
$MANPATH = /usr/share/man:/usr/local/share/man:/Applications/Ghostty.app/Contents/Resources/ghostty/../man:
```

**Comment by krobelus** [comment]
> With completely unconfigured zsh, I get:

Confirmed that I get the same result when running `path_helper` from fish

```
> MANPATH=:/Applications/Ghostty.app/Contents/Resources/ghostty/../man /usr/libexec/path_helper | grep MANPATH
MANPATH="/usr/share/man:/usr/local/share/man:/Applications/Ghostty.app/Contents/Resources/ghostty/../man:"; export MANPATH;
```

So the only difference is that we're putting the second colon first instead of last

**Comment by krobelus** [comment]
ok I think I'm close to a fix for the extra middle colon, but where does the extra colon at the end come from?
https://raw.githubusercontent.com/apple-opensource-mirror/shell_cmds/4ea11424b80a117ad1e401867b1453bb2362feb3/path_helper/path_helper.c
doesn't seem to set it

Given MANPATH=:/my/humble/manpath, the last paragraph of construct_path should execute as

result = /usr/share/man:/usr/local/share/man
str = :/my/humble/manpath
str = \0/my/humble/manpath
append_path_segment(&result, str); // no-op
str = /my/humble/manpath
append_path_segment(&result, str);
result = /usr/share/man:/usr/local/share/man:/my/humble/manpath

**Comment by injust** [comment]
> ok I think I'm close to a fix for the extra middle colon, but where does the extra colon at the end come from? [apple-opensource-mirror/shell_cmds@`4ea1142`/path_helper/path_helper.c (raw)](https://raw.githubusercontent.com/apple-opensource-mirror/shell_cmds/4ea11424b80a117ad1e401867b1453bb2362feb3/path_helper/path_helper.c) doesn't seem to set it

I've been wondering the exact same thing...

**Comment by injust** [comment]
When I compile the code myself, the output is different from the `path_helper` that ships with macOS...

```fish
$ MANPATH= ./a.out | grep MANPATH
MANPATH="/usr/share/man:/usr/local/share/man"; export MANPATH;

$ MANPATH= /usr/libexec/path_helper | grep MANPATH
MANPATH="/usr/share/man:/usr/local/share/man:"; export MANPATH;
```

**Comment by injust** [comment]
Seems like https://raw.githubusercontent.com/apple-opensource-mirror/shell_cmds/4ea11424b80a117ad1e401867b1453bb2362feb3/path_helper/path_helper.c is outdated.

https://github.com/apple-oss-distributions/shell_cmds/blob/main/path_helper/path_helper.c gives me behaviour matching the `/usr/libexec/path_helper` shipped with macOS. It explicitly adds the trailing colon in `main()`.

**Comment by krobelus** [comment]
nice, that's certainly easier than trying to attach a debugger ;)
Makes me wonder why we don't use `path_helper` directly..
We could do `eval "$(/usr/libexec/path_helper -c)"` since we have a compatibility wrapper for csh's `setenv`.
But `path_helper` doesn't do any quoting, so quotes or dollar signs in PATH could inject code.. so I guess we either do some sanitization or stick to our reimplementation until Apple adds `path_helper -f`

**Comment by injust** [comment]
The more I dig into this, the more I feel like `path_helper` is a PITA...

It turns `MANPATH=:foo` (something that is _appended_ to man's default search path) into `MANPATH=/usr/share/man:/usr/local/share/man:foo:` (which is now _prepended_ to man's search path).

The prepend basically blows away anything that comes from `man`'s default search path (which takes your $PATH into consideration).

😭

**Comment by krobelus** [comment]
> it basically blows away anything that comes from man's default search path

Right, and surprisingly the default search path has some components interpsersed between `/usr/local/share/man` (which is the 1st entry below) and `/usr/share/man` (the 3rd), God knows why 🤷 

```
$ manpath
/usr/local/share/man:/System/Cryptexes/App/usr/share/man:/usr/share/man:/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/share/man:/Library/Developer/CommandLineTools/usr/share/man
```

> which takes your $PATH into consideration

not sure how

**Comment by krobelus** [comment]
I pushed a "fix" to the `faithful-macos-manpath` branch.
Not sure if we should take it though,
it matches macOS behavior but it seems just wrong.

fish's current behavior, i.e. 

```
MANPATH=:/Applications/Ghostty.app/Contents/Resources/ghostty/../man fish -lc 'echo $MANPATH'
/usr/share/man:/usr/local/share/man::/Applications/Ghostty.app/Contents/Resources/ghostty/../man
```

seems exactly what the user intended to happen - the system default manpath is inserted before the Ghostty one.
(God knows why `/usr/share/man` etc. must be prepended(!) explicitly rather than leaving this to the default manpath..)

**Comment by injust** [comment]
> > it basically blows away anything that comes from man's default search path
> 
> Right, and surprisingly the default search path has some components interpsersed between `/usr/local/share/man` (which is the 1st entry below) and `/usr/share/man` (the 3rd), God knows why 🤷
> 
> ```
> $ manpath
> /usr/local/share/man:/System/Cryptexes/App/usr/share/man:/usr/share/man:/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/share/man:/Library/Developer/CommandLineTools/usr/share/man
> ```
> 
> > which takes your $PATH into consideration
> 
> not sure how

The `manpath` man page:

```
IMPLEMENTATION NOTES
     The manpath utility constructs the manual path from two sources:
     1.   From each component of the user's PATH for the first of:
          -   pathname/man
          -   pathname/MAN
          -   If pathname ends with /bin: pathname/../share/man and pathname/../man
     2.   The configuration files listed in the FILES section for MANPATH entries.
     The information from these locations is then concatenated together.

     If the -L flag is set, the manpath utility will search the configuration files listed in the FILES section for MANLOCALE entries.
```

**Comment by injust** [comment]
> I pushed a "fix" to the `faithful-macos-manpath` branch. Not sure if we should take it though, it matches macOS behavior but it seems just wrong.
> 
> fish's current behavior, i.e.
> 
> ```
> MANPATH=:/Applications/Ghostty.app/Contents/Resources/ghostty/../man fish -lc 'echo $MANPATH'
> /usr/share/man:/usr/local/share/man::/Applications/Ghostty.app/Contents/Resources/ghostty/../man
> ```
> 
> seems exactly what the user intended to happen - the system default manpath is inserted before the Ghostty one. (God knows why `/usr/share/man` etc. must be prepended(!) explicitly rather than leaving this to the default manpath..)

Agreed, it seems incorrect that `path_helper` just discards empty segments when parsing the existing $MANPATH.

**Comment by injust** [comment]
I think fish should be faithful to macOS and all its quirks. You'd run into the same thing if you used zsh, which actually invokes `path_helper`.

_That being said_, I hate how `path_helper` prepends its own defaults, so in my own dotfiles I ended up removing any inherited $MANPATH (with `set -ge MANPATH`) to make sure `path_helper` doesn't add its defaults.

**Comment by krobelus** [comment]
> set -ge MANPATH

how do you do that though? `share/config.fish` which runs `__fish_macos_set_env` is sourced before user config,
so the only way to disable it would be to override the `__fish_macos_set_env` function (which would only work on the above branch).
`__fish_macos_set_env` is only run in login shells; on macOS, every terminal tab creates a login shell AFAICT

So I think this works only if you have nested fish where the inner one is a login shell. (Or when using `path_helper` directly).

I guess ghostty exports MANPATH to the child shell.

Until Apple fixes this, I wonder if users/apps should work around this issue by never setting `MANPATH` -- because then `path_helper` will not touch it.
Ghostty would need to find another way to inject `MANPATH`. There must be one, given the output of `manpath` on my system

**Comment by injust** [comment]
> how do you do that though? `share/config.fish` which runs `__fish_macos_set_env` is sourced before user config

I put this in my `$__fish_config_dir/conf.d/manpath.fish` (which runs before `share/config.fish`):

```fish
if status --is-login; and command -q /usr/libexec/path_helper
    # Remove inherited $MANPATH to stop macOS' path_helper from prepending to man's standard search path (https://github.com/fish-shell/fish-shell/issues/10684)
    set -ge MANPATH
end
```

It removes Ghostty from $MANPATH, but everything still works correctly for me because Homebrew installs Ghostty's man pages:

https://github.com/Homebrew/homebrew-cask/blob/6876bcf1373119f25a1d6661e0bc722cf3bb57fb/Casks/g/ghostty.rb#L20-L21

> Until Apple fixes this

I'm not sure I'd count on that happening 😆

> I wonder if users/apps should work around this issue by never setting `MANPATH` -- because then `path_helper` will not touch it. Ghostty would need to find another way to inject `MANPATH`.

That's fair -- I only ran into this issue because Ghostty was appending itself to $MANPATH (which is a totally correct thing for them to do, but it just happened to reveal this `path_helper` behaviour).

But this quirk is macOS-exclusive ($MANPATH should work properly on Linux). I also don't think many users care about $MANPATH correctness that much...

> There must be one, given the output of `manpath` on my system

I think most of the entries in `manpath` are from man's default search path (which searches directories in your $PATH). Look at the ENVIRONMENT section of `man 1 man` and the IMPLEMENTATION NOTES section of `man 1 manpath`.
